File: lib_rxdownloader/src/main/java/com/hardy/rx/downloader/DownloadGroup.java
package com.changba.plugin.livechorus.download;

import com.changba.library.commonUtils.KTVLog;
import com.changba.plugin.livechorus.download.model.ProgressModel;

import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.RejectedExecutionHandler;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.locks.ReentrantReadWriteLock;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;

/**
 * @author Hdq on 2021/3/9.
 * 可参考：https://www.cnblogs.com/baiqiantao/p/10679677.html
 * https://github.com/lingochamp/FileDownloader/blob/master/README-zh.md
 */
public class DownloadGroup {
    private static final String TAG = "----DownloadGroup";

    private Subject<DownloadTask> downloadTaskSubject;
    private LinkedList<DownloadTask> waitingTasks = new LinkedList<>();
    private LinkedList<DownloadTask> downloadingTasks = new LinkedList<>();
    private List<DownloadTask> allTasks = new ArrayList<>();

    private Subscription downloadQueueSubscription;
    private CompositeDisposable mSubscriptions = new CompositeDisposable();

    private int limit;

    private ThreadPoolExecutor executor;

    private int count;

    private ReentrantReadWriteLock lock = new ReentrantReadWriteLock();

    private AtomicBoolean isStarted = new AtomicBoolean(false);
    private AtomicBoolean isDestroy = new AtomicBoolean(false);
    private Object startStopLock = new Object();

    /**
     *
     * @param limit 最大并发下载数量
     */
    public DownloadGroup(int limit) {
        this.limit = limit;
        Collections.synchronizedList(waitingTasks);
        Collections.synchronizedList(downloadingTasks);
        executor = new ThreadPoolExecutor(0, limit * 2, 1000 * 60, TimeUnit.MILLISECONDS,
                new SynchronousQueue<>(), new NamedThreadFactory("DownloadGroup"), new RejectedExecutionHandler() {
            @Override
            public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
                throw new RejectedExecutionException("Task " + r.toString() +
                        " rejected from " +
                        executor.toString() + ",无法下载，超出最大线程限制");
            }
        });
        downloadTaskSubject = PublishSubject.<DownloadTask>create().toSerialized();
    }

    public void start() {
        if (isStarted.get()) { return; }
        synchronized (startStopLock) {
            isStarted.getAndSet(true);
            isDestroy.getAndSet(false);
            downloadTaskSubject.toFlowable(BackpressureStrategy.DROP)
                    .observeOn(Schedulers.io(), false, 10000)
                    .subscribe(new Subscriber<DownloadTask>() {

                        @Override
                        public void onSubscribe(Subscription s) {
                            downloadQueueSubscription = s;
                            s.request(limit);
                        }

                        @Override
                        public void onError(Throwable e) {
                            e.printStackTrace();
                        }

                        @Override
                        public void onComplete() {

                        }

                        @Override
                        public void onNext(DownloadTask downloadTask) {
                            count++;
                            KTVLog.d(TAG, "onNext: 从队列中取出一个下载任务 count = " + count);
                            try {
                                lock.writeLock().lock();
                                waitingTasks.remove(downloadTask);
                                downloadingTasks.add(downloadTask);
                            } catch (Exception e) {
                                e.printStackTrace();
                            } finally {
                                lock.writeLock().unlock();
                            }
                            mSubscriptions.add(downloadTask.getProgressObservable().subscribe(
                                    progress -> {//onNext
                                    }, throwable -> {//onError
                                        throwable.printStackTrace();
                                    }, () -> {//onCompleted
                                        KTVLog.d(TAG, "下载完成onCompleted: ");
                                        if (downloadQueueSubscription != null) {
                                            downloadQueueSubscription.request(1);
                                        }
                                    }
                            ));

                            try {
                                KTVLog.d(TAG, "onNext: 开线程执行下载");
                                executor.execute(() -> {
                                    KTVLog.d(TAG, "downloadTask.start - thread name=" + Thread.currentThread());
                                    try {
                                        downloadTask.startSync();
                                    } catch (Exception e) {
                                        e.printStackTrace();
                                    }
                                });
                            } catch (Exception e) {
                                KTVLog.e(TAG, "线程池错误", e);
                                downloadTask.notifyError(e);
                            }
                        }
                    });
        }
    }

    /**
     * 停止下载
     * 只是停止，重新调用start还会继续之前的下载
     */
    public void stopGroup() {
        synchronized (startStopLock) {
            if (downloadQueueSubscription != null) {
                downloadQueueSubscription.cancel();
            }
            mSubscriptions.clear();
            try {
                lock.writeLock().lock();
                Iterator<DownloadTask> iterator = downloadingTasks.iterator();
                while (iterator.hasNext()) {
                    DownloadTask task  = iterator.next();
                    task.stop();
                    if (task.getState() != DownloadTask.State.DOWNLOAD_SUCCESS) {
                        iterator.remove();
                        waitingTasks.add(task);//放回waitingTasks
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                lock.writeLock().unlock();
            }
            isStarted.getAndSet(false);
        }
    }

    /**
     * 停止下载
     * 清空下载任务
     */
    public void destroyGroup() {
        synchronized (startStopLock) {
            try {
                lock.writeLock().lock();
                stopGroup();
                waitingTasks.clear();
                downloadingTasks.clear();
                allTasks.clear();
                isDestroy.getAndSet(true);
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                lock.writeLock().unlock();
            }
        }
    }

    public DownloadTask enqueue(String downloadUrl, String localFileName) {
        return enqueue(downloadUrl,localFileName,1);
    }

    public DownloadTask enqueue(String downloadUrl, String localFileName, float weight) {
        return enqueue(downloadUrl,localFileName,weight,false);
    }

    public DownloadTask enqueue(String downloadUrl, String localFileName, boolean isOnlyCheckLocalFileExit) {
        return enqueue(downloadUrl,localFileName,1,isOnlyCheckLocalFileExit);
    }

    public DownloadTask enqueue(String downloadUrl, String localFileName, float weight, boolean isOnlyCheckLocalFileExit) {
        DownloadTask task = findTask(downloadUrl);
        if (task != null && task.getState() == DownloadTask.State.DOWNLOAD_SUCCESS && !new File(localFileName).exists()) {
            //处理异常情况，这种情况发生在：app开着，手动删除本地文件
            downloadingTasks.remove(task);
            allTasks.remove(task);
            task = null;
        }
        if (task == null) {
            task = new DownloadTask(downloadUrl,localFileName,weight);
            task.setOnlyCheckLocalFileExit(isOnlyCheckLocalFileExit);
            try {
                lock.writeLock().lock();
                waitingTasks.add(task);
                allTasks.add(task);
                isReComputeTotalWeight.getAndSet(true);
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                lock.writeLock().unlock();
            }
            downloadTaskSubject.onNext(task);
            KTVLog.d(TAG, "enqueue: downloadUrl=" + downloadUrl + ",localFileName=" + localFileName);
        } else {
            task.setOnlyCheckLocalFileExit(isOnlyCheckLocalFileExit);
            KTVLog.w(TAG, "enqueue: 下载任务已存在 downloadUrl=" + downloadUrl + ",localFileName=" + localFileName);
        }
        return task;
    }


    public DownloadTask findTask(String url) {
        try {
            lock.readLock().lock();
            Iterator<DownloadTask> iterator = allTasks.iterator();
            while (iterator.hasNext()) {
                DownloadTask task = iterator.next();
                if (task.getDownloadUrl().equals(url)) {
                    return task;
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            lock.readLock().unlock();
        }
        return null;
    }

    public List<DownloadTask> getWaitingTasks() {
        return waitingTasks;
    }

    public List<DownloadTask> getDownloadingTasks() {
        return downloadingTasks;
    }

    public List<DownloadTask> getAllTasks() {
        List<DownloadTask> tempList = new ArrayList<>();
        try {
            lock.readLock().lock();
            tempList.addAll(allTasks);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            lock.readLock().unlock();
        }
        return tempList;
    }


    public float getTotalProgress(){
        float totalWeight = getTotalWeight();
        float avgProgress = 0;
        try {
            lock.readLock().lock();
            Iterator<DownloadTask> iterable = allTasks.iterator();
            while (iterable.hasNext()) {
                DownloadTask task = iterable.next();
                avgProgress += task.getProgress() * (task.getWeight() / totalWeight);
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            lock.readLock().unlock();
        }
        KTVLog.d(TAG, "progress-----1: " + avgProgress);
        return avgProgress;
    }

    private AtomicBoolean isReComputeTotalWeight = new AtomicBoolean(false);
    private float totalWeight = 0;
    public float getTotalWeight() {
        if (isReComputeTotalWeight.get()) {
            totalWeight = 0;
            try {
                lock.readLock().lock();
                Iterator<DownloadTask> iterable = allTasks.iterator();
                while (iterable.hasNext()) {
                    DownloadTask task = iterable.next();
                    totalWeight += task.getWeight();
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                lock.readLock().unlock();
            }
            isReComputeTotalWeight.getAndSet(false);
            KTVLog.d(TAG, "getTotalWeight: " + totalWeight);
        }
        return totalWeight;
    }

    /***
     * 所有任务的总进度
     * @return
     */
    public Observable<ProgressModel> getTotalProgressObservable(){
        List<Observable<ProgressModel>> allProgressObservable = new ArrayList<>();
        ProgressModel allProgressModel = new ProgressModel();
        try {
            lock.readLock().lock();
            Iterator<DownloadTask> iterable = allTasks.iterator();
            while (iterable.hasNext()) {
                DownloadTask task = iterable.next();
                allProgressObservable.add(task.getProgressObservable());
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            lock.readLock().unlock();
        }
        return Observable.merge(allProgressObservable)
                .sample(200,TimeUnit.MILLISECONDS)
                .map(progressModel -> {
                    allProgressModel.setTask(progressModel.getTask());
                    allProgressModel.setProgress(getTotalProgress());
                    return allProgressModel;
                });
    }

    public void testReset(){
        allTasks.clear();
        downloadingTasks.clear();
        waitingTasks.clear();
    }

    /**
     * 当前的所有任务下载完成
     * @return
     */
    public boolean isAllTaskDownloadSuccess() {
        List<DownloadTask> tasks = getAllTasks();
        for (DownloadTask task : tasks) {
            if (task.getState() != DownloadTask.State.DOWNLOAD_SUCCESS) {
                return false;
            }
        }
        return true;
    }

    public boolean isDestroy() {
        return isDestroy.get();
    }
}

File: lib_rxdownloader/src/main/java/com/wuzhu/rx/downloader/DownloadGroup.kt
package com.wuzhu.rx.downloader

import android.util.Log
import com.wuzhu.rx.downloader.model.*
import io.reactivex.BackpressureOverflowStrategy
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.ReplaySubject
import org.reactivestreams.Subscriber
import org.reactivestreams.Subscription
import java.io.File
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.SynchronousQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

private const val TAG = "----DownloadGroup"

/**
 * @author Hdq on 2021/3/9.
 * 可参考：https://www.cnblogs.com/baiqiantao/p/10679677.html
 * https://github.com/lingochamp/FileDownloader/blob/master/README-zh.md
 * @param limit 最大并发下载数量
 */
class DownloadGroup(private val limit: Int) {
    private val business by lazy { DownloadGroupBusiness() }

    private val progressSubject = BehaviorSubject.create<GroupProgressModel>().toSerialized()

    /**
     * 这里提前new出来，避免使用时候频繁创建对象
     * 范围 0 - 100
     */
    private val progressModel = GroupProgressModel(this)
    private val stateSubject = BehaviorSubject.create<GroupStateModel>().toSerialized()

    /**
     * 这里提前new出来，避免使用时候频繁创建对象
     */
    private val groupStateModel = GroupStateModel(this)

    private val downloadTaskSubject = ReplaySubject.create<DownloadTask>().toSerialized()
    private var downloadQueueSubscription: Subscription? = null

    private val executor: ThreadPoolExecutor = ThreadPoolExecutor(
        0, limit * 2, 1000 * 60, TimeUnit.MILLISECONDS, SynchronousQueue(), NamedThreadFactory("DownloadGroup")
    ) { r: Runnable, executor: ThreadPoolExecutor ->
        throw RejectedExecutionException(
            "Task $r rejected from $executor,无法下载，超出最大线程限制"
        )
    }
    private var count = 0

    @Volatile
    private var isCreated = false

    @Volatile
    private var isStarted = false

    private val lifecycleLock = Object()

    init {
        groupStateModel.state = State.PREPARE
        stateSubject.onNext(groupStateModel)
    }


    private fun produce(task: DownloadTask) {
        Log.w(TAG, "produce: $task")
        business.waitingQueue.add(task)
        downloadTaskSubject.onNext(task)
    }

    private fun consume(task: DownloadTask) {
        synchronized(lifecycleLock) {
            if (!isStarted) {
                lifecycleLock.wait() //阻断消费
            }
            if (task == business.emptyTask) {
                Log.e(TAG, "收到空任务，说明有异常")
                return
            }
            count++
            business.downloadingTasks.add(task)
            startNextDownloadTask(task)
            business.waitingQueue.remove(task)
            Log.w(TAG, "consume: $task , count = $count")
        }
    }

    private fun startNextDownloadTask(downloadTask: DownloadTask) {
        try {
            executor.execute {
                downloadTask.startSync()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            downloadTask.notifyError(e)
        }
    }

    private fun registerDownloadState(downloadTask: DownloadTask) {
        downloadTask.getStateObservable().subscribe(object : DisposableObserver<TaskStateModel>() {
            override fun onNext(t: TaskStateModel) {
                handleDownloadState(t)
            }

            override fun onError(e: Throwable) {
                e.printStackTrace()
            }

            override fun onComplete() {
                Log.e(TAG, "DownloadState--onCompleted: ")
            }
        })

    }

    private fun registerDownloadProgress(downloadTask: DownloadTask) {
        downloadTask.getProgressObservable().sample(10, TimeUnit.MILLISECONDS,true).subscribe(object : DisposableObserver<TaskProgressModel>() {
            override fun onNext(t: TaskProgressModel) {
                progressModel.progress = business.getGroupProgress()
                progressSubject.onNext(progressModel)
            }

            override fun onError(e: Throwable) {
                e.printStackTrace()
            }

            override fun onComplete() {
                Log.e(TAG, "DownloadProgress--onComplete: ")
            }
        })
    }

    private fun handleDownloadState(taskStateModel: TaskStateModel) {
        when (taskStateModel.state) {
            State.PREPARE -> {}
            State.START,State.DOWNLOADING -> {
                groupStateModel.state = taskStateModel.state
                stateSubject.onNext(groupStateModel)
            }
            State.SUCCESS, State.STOPPED, State.ERROR -> {
                business.downloadingTasks.remove(taskStateModel.task)
                downloadQueueSubscription?.request(1)
                if (business.downloadingTasks.isEmpty()) {
                    groupStateModel.state = business.getGroupState()
                    Log.w(TAG, "handleDownloadState: group下载结束，通知下载状态::state=$groupStateModel")
                    stateSubject.onNext(groupStateModel)
                }
            }
        }
    }


    fun create() {
        Log.w(TAG, "create: $this")
        synchronized(lifecycleLock) {
            if (isCreated) {
                return
            }
            downloadTaskSubject.toFlowable(BackpressureStrategy.BUFFER).onBackpressureBuffer(Long.MAX_VALUE, {
                    Log.e(TAG, "背压队列溢出，删除最新的数据::超过能处理的最大下载量，会丢失下载任务")
                }, BackpressureOverflowStrategy.DROP_LATEST).onErrorReturn {
                    it.printStackTrace()
                    business.emptyTask
                }.observeOn(Schedulers.io(), false, Flowable.bufferSize() * 2)//buffer是背压缓存队列大小
                .subscribe(object : Subscriber<DownloadTask> {
                    override fun onSubscribe(s: Subscription) {
                        downloadQueueSubscription = s
                        s.request(limit.toLong())
                    }

                    override fun onError(e: Throwable) {
                        e.printStackTrace()
                    }

                    override fun onComplete() {}
                    override fun onNext(downloadTask: DownloadTask) {
                        consume(downloadTask)
                    }
                })
            isCreated = (true)
        }
        Log.w(TAG, "created: $this")
    }

    fun start() {
        Log.w(TAG, "start: $this")
        synchronized(lifecycleLock) {
            if (isStarted) {
                return
            }
            //判断任务是否需要重新下载
            business.allTasks.forEach {
                reDownloadTask(it)
            }
            isStarted = (true)
            lifecycleLock.notify()//通知开始消费
        }
        Log.w(TAG, "started: $this")
    }

    /**
     * 停止下载
     * 只是停止，重新调用start还会继续之前的下载
     */
    fun stop() {
        Log.w(TAG, "stop: $this")
        synchronized(lifecycleLock) {
            //先阻断消费
            isStarted = (false)
            business.downloadingTasks.forEach {
                //停掉任务
                it.stop()
            }
        }
        Log.w(TAG, "stoped: $this")
    }

    /**
     * 停止下载
     * 清空下载任务
     */
    fun destroy() {
        Log.w(TAG, "destroy: $this")
        synchronized(lifecycleLock) {
            stop()
            downloadQueueSubscription?.cancel()
            business.destroyAllTask()
            isCreated = (false)
        }
        Log.w(TAG, "destroyed: $this")
    }

    @JvmOverloads
    fun addTask(
        downloadUrl: String, localFileName: String, weight: Float = 1f, isOnlyCheckLocalFileExit: Boolean = false
    ): DownloadTask {
        synchronized(lifecycleLock) {
            var task = findTask(downloadUrl)
            if (task == null) {
                Log.w(TAG, "addTask::任务不存在准备新建下载任务")
                task = createTask(downloadUrl, localFileName, weight, isOnlyCheckLocalFileExit)
                produce(task)
            } else {
                task.isOnlyCheckLocalFileExit = isOnlyCheckLocalFileExit
                Log.w(TAG, "addTask::下载任务已存在，进入reDownloadTask，判断是否需要重新下载")
                reDownloadTask(task)
            }
            return task
        }
    }

    /***
     * 根据状态判断任务是否需要重新下载
     */
    private fun reDownloadTask(task: DownloadTask) {
        when (task.getState()) {
            State.PREPARE, State.START, State.DOWNLOADING -> {
                Log.w(TAG, "reDownloadTask:在下载流程中，无需重新下载- $task")
            }
            State.SUCCESS -> {
                //处理异常情况，状态成功，文件不存在
                //这种情况发生在：app开着，手动删除本地文件
                if (!File(task.localFileName).exists()) {
                    Log.w(TAG, "reDownloadTask:触发重新下载::本地文件被删除- $task")
                    task.reset(true)
                    produce(task)
                } else {
                    Log.w(TAG, "reDownloadTask:已下载成功，无需重新下载- $task")
                }
            }
            State.STOPPED, State.ERROR -> {
                //重新下载任务
                Log.w(TAG, "reDownloadTask:触发重新下载-- $task")
                task.reset(!task.isTempFileExists())//STOPPED 和 ERROR 也可能删除了本地文件
                produce(task)
            }
            else -> {
                Log.w(TAG, "reDownloadTask:other，无需重新下载- $task")
            }
        }
    }


    private fun createTask(
        downloadUrl: String, localFileName: String, weight: Float = 1f, isOnlyCheckLocalFileExit: Boolean = false
    ): DownloadTask {
        val task = DownloadTask(downloadUrl, localFileName, weight)
        task.isOnlyCheckLocalFileExit = isOnlyCheckLocalFileExit
        business.allTasks.add(task)
        business.isReComputeTotalWeight.getAndSet(true)
        //订阅进度 - 只订阅一次
        registerDownloadProgress(task)
        //订阅下载状态 - 只订阅一次
        registerDownloadState(task)
        Log.w(TAG, "createTask::新建下载任务 $task")
        return task
    }

    fun findTask(url: String): DownloadTask? = business.findTask(url)

    fun getProgressObservable(): Flowable<GroupProgressModel> = progressSubject.toFlowable(BackpressureStrategy.LATEST).onBackpressureBuffer(10000, {
            Log.e(TAG, "背压队列溢出，删除最旧的数据")
        }, BackpressureOverflowStrategy.DROP_OLDEST).sample(100, TimeUnit.MILLISECONDS,true)

    fun getProgress() = progressModel.progress

    fun getStateObservable(): Flowable<GroupStateModel> = stateSubject.toFlowable(BackpressureStrategy.LATEST).onBackpressureBuffer(10000, {
            Log.e(TAG, "背压队列溢出，删除最旧的数据")
        }, BackpressureOverflowStrategy.DROP_OLDEST)

    fun getState() = groupStateModel.state

    override fun toString(): String {
        return "DownloadGroup::${this.hashCode()}(isCreated=$isCreated, isStarted=$isStarted)"
    }


}